<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scDblFinder">
<title>Detecting clusters of doublet cells with DE analyses • scDblFinder</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Detecting clusters of doublet cells with DE analyses">
<meta property="og:description" content="scDblFinder">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scDblFinder</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.15.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/scDblFinder.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/computeDoubletDensity.html">Scoring potential doublets from simulated densities</a>
    <a class="dropdown-item" href="../articles/findDoubletClusters.html">Detecting clusters of doublet cells with DE analyses</a>
    <a class="dropdown-item" href="../articles/introduction.html">Introduction to the scDblFinder package</a>
    <a class="dropdown-item" href="../articles/recoverDoublets.html">Recovering intra-sample doublets</a>
    <a class="dropdown-item" href="../articles/scATAC.html">Doublet identifiation in single-cell ATAC-seq</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/plger/scDblFinder/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Detecting clusters of doublet cells with DE analyses</h1>
                        <h4 data-toc-skip class="author">Aaron Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                  
            <h4 data-toc-skip class="date">2023-09-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/plger/scDblFinder/blob/HEAD/vignettes/findDoubletClusters.Rmd" class="external-link"><code>vignettes/findDoubletClusters.Rmd</code></a></small>
      <div class="d-none name"><code>findDoubletClusters.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="tldr">tl;dr<a class="anchor" aria-label="anchor" href="#tldr"></a>
</h2>
<p>To demonstrate, we’ll use one of the mammary gland datasets from the
<em><a href="https://bioconductor.org/packages/3.17/scRNAseq" class="external-link">scRNAseq</a></em>
package. We will subset it down to a random set of 500 cells for
speed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/BachMammaryData.html" class="external-link">BachMammaryData</a></span><span class="op">(</span>samples<span class="op">=</span><span class="st">"G_2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fl">500</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>For the purposes of this demonstration, we’ll perform an extremely
expedited analysis. One would usually take more care here and do some
quality control, create some diagnostic plots, etc., but we don’t have
the space for that.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents<span class="op">=</span><span class="fl">10</span>, subset_row<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n<span class="op">=</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span>, text_by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="findDoubletClusters_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>We then run <code><a href="../reference/findDoubletClusters.html">findDoubletClusters()</a></code> to test each cluster
against the null hypothesis that it <em>does</em> consist of doublets.
The null is rejected if a cluster has many DE genes that lie outside the
expression limits defined by the “source” clusters. On the other hand,
if <code>num.de</code> is low, the cluster’s expression profile is
consistent with the doublet hypothesis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/plger/scDblFinder" class="external-link">scDblFinder</a></span><span class="op">)</span></span>
<span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findDoubletClusters.html">findDoubletClusters</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">clusters</span><span class="op">)</span></span>
<span><span class="va">tab</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 7 rows and 9 columns</span></span>
<span><span class="co">##       source1     source2    num.de median.de               best     p.value</span></span>
<span><span class="co">##   &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;        &lt;character&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 6           2           1         0       122 ENSMUSG00000019256 1.00000e+00</span></span>
<span><span class="co">## 3           6           5         8        46 ENSMUSG00000002985 8.41352e-05</span></span>
<span><span class="co">## 7           5           1        11       393 ENSMUSG00000075705 2.04690e-10</span></span>
<span><span class="co">## 4           7           2        24        63 ENSMUSG00000022491 4.15645e-13</span></span>
<span><span class="co">## 1           7           6        86       715 ENSMUSG00000001349 8.49703e-08</span></span>
<span><span class="co">## 5           7           2       112      1007 ENSMUSG00000024610 2.38765e-11</span></span>
<span><span class="co">## 2           6           5       124       336 ENSMUSG00000023505 4.06936e-06</span></span>
<span><span class="co">##   lib.size1 lib.size2      prop</span></span>
<span><span class="co">##   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 6  0.927081  0.443667     0.084</span></span>
<span><span class="co">## 3  2.026090  0.859424     0.242</span></span>
<span><span class="co">## 7  1.195040  1.249946     0.030</span></span>
<span><span class="co">## 4  0.630479  1.646729     0.150</span></span>
<span><span class="co">## 1  0.800035  2.253942     0.192</span></span>
<span><span class="co">## 5  0.836792  2.185591     0.032</span></span>
<span><span class="co">## 2  1.078655  0.457542     0.270</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="mathematical-background">Mathematical background<a class="anchor" aria-label="anchor" href="#mathematical-background"></a>
</h2>
<p>Consider a cell population <span class="math inline">\(i\)</span>
that has mean transcript count <span class="math inline">\(\lambda_{gi}\)</span> for gene <span class="math inline">\(g\)</span>. Assume that each population exhibits a
unique scaling bias <span class="math inline">\(s_i\)</span>,
representing the efficiency of library preparation for that population.
The observed read/UMI count for each gene is then <span class="math inline">\(\mu_{gi}=s_i\lambda_{gi}\)</span>. (For
simplicity, we will ignore gene-specific scaling biases, as this is
easily accommodated by considering <span class="math inline">\(\lambda_{gi} \equiv \phi_g \lambda_{gi}\)</span>
for some bias <span class="math inline">\(\phi_g\)</span>.) The expected
total count for each population is <span class="math inline">\(N_i =
\sum_g \mu_{gi}\)</span>.</p>
<p>Now, let us consider a doublet population <span class="math inline">\(j\)</span> that forms from two parent populations
<span class="math inline">\(i_1\)</span> and <span class="math inline">\(i_2\)</span>. The observed read count for <span class="math inline">\(g\)</span> in <span class="math inline">\(j\)</span> is <span class="math inline">\(\mu_{gj}
= s_j (\lambda_{gi_1} + \lambda_{gi_2})\)</span>. Note that <span class="math inline">\(s_j\)</span> need not be any particular function
of <span class="math inline">\(s_{i_1}\)</span> and <span class="math inline">\(s_{i_2}\)</span>. Rather, this relationship
depends on how quickly the reverse transcription and amplification
reagents are saturated during library preparation, which is difficult to
make assumptions around.</p>
</div>
<div class="section level2">
<h2 id="normalization-by-library-size">Normalization by library size<a class="anchor" aria-label="anchor" href="#normalization-by-library-size"></a>
</h2>
<p>We obtain log-normalized expression values for each cell based on the
library size. Assume that the library size-normalized expression values
are such that <span class="math inline">\(\mu_{gi_1}N_{i_1}^{-1} &lt;
\mu_{gi_2}N_{i_2}^{-1}\)</span>, i.e., the proportion of <span class="math inline">\(g\)</span> increases in <span class="math inline">\(i_2\)</span> compared to <span class="math inline">\(i_1\)</span>. The contribution of each <span class="math inline">\(s_i\)</span> cancels out, yielding <span class="math display">\[
\frac{\lambda_{gi_1}}{\sum_g \lambda_{gi_1}} &lt;
\frac{\lambda_{gi_2}}{\sum_g \lambda_{gi_2}} \;.
\]</span> The normalized expression value of the doublet cluster <span class="math inline">\(j\)</span> is subsequently <span class="math display">\[
\frac{\lambda_{gi_1} + \lambda_{gi_2}}{\sum_g (\lambda_{gi_1} +
\lambda_{gi_2})} \;,
\]</span> and it is fairly easy to show that <span class="math display">\[
\frac{\lambda_{gi_1}}{\sum_g \lambda_{gi_1}} &lt;
\frac{\lambda_{gi_1} + \lambda_{gi_2}}{\sum_g (\lambda_{gi_1} +
\lambda_{gi_2})}  &lt;
\frac{\lambda_{gi_2}}{\sum_g \lambda_{gi_2}} \;.
\]</span> In other words, the expected library size-normalized
expression of our gene in the doublet cluster lies between that of the
two parents.</p>
<p>It is harder to provide theoretical guarantees with arbitrary size
factors, which is why we only use the library sizes for normalization
instead. The exception is that of spike-in size factors that would
estimate <span class="math inline">\(s_i\)</span> directly. This would
allow us to obtain estimates of <span class="math inline">\(\lambda_{gi}\)</span> for the parent clusters and
of <span class="math inline">\(\lambda_{gi_1} + \lambda_{gi_2}\)</span>
for the doublets. In this manner, we could more precisely identify
doublet clusters as those where the normalized expression value is equal
to the sum of the parents. Unfortunately, spike-ins are generally not
available for droplet-based data sets where doublets are most
problematic.</p>
</div>
<div class="section level2">
<h2 id="testing-for-lack-of-intermediacy">Testing for (lack of) intermediacy<a class="anchor" aria-label="anchor" href="#testing-for-lack-of-intermediacy"></a>
</h2>
<p>We want to identify the clusters that may be comprised of doublets of
other clusters. For each cluster <span class="math inline">\(j'\)</span>, we test for differential
expression in the library size-normalized expression profiles against
every other cluster <span class="math inline">\(i'\)</span>. For
each pair of other clusters <span class="math inline">\(i'_1\)</span> and <span class="math inline">\(i'_2\)</span>, we identify genes that change
in <span class="math inline">\(j'\)</span> against both <span class="math inline">\(i'_1\)</span> and <span class="math inline">\(i'_2\)</span> <strong>in the same
direction</strong>. The presence of such genes violates the intermediacy
expected of a doublet cluster and provides evidence that <span class="math inline">\(j'\)</span> is not a doublet of <span class="math inline">\(i'_1\)</span> and <span class="math inline">\(i'_2\)</span>.</p>
<p>Significant genes are identified by an intersection-union test on the
<span class="math inline">\(p\)</span>-values from the pairwise
comparisons between <span class="math inline">\(j'\)</span> and
<span class="math inline">\(i'_1\)</span> or <span class="math inline">\(i'_2\)</span>. (Specifically, <span class="math inline">\(t\)</span>-tests are used via the
<code><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers()</a></code> function from <em><a href="https://bioconductor.org/packages/3.17/scran" class="external-link">scran</a></em>.) The
<span class="math inline">\(p\)</span>-value for a gene is set to unity
when the signs of the log-fold changes are not the same between
comparisons. Multiple correction testing is applied using the
Benjamini-Hochberg method, and the number of genes detected at a
specified false discovery rate (usually 5%) is counted. The pair <span class="math inline">\((i'_1, i'_2)\)</span> with the fewest
detected genes are considered as the putative parents of <span class="math inline">\(j'\)</span>.</p>
<p>In theory, it is possible to compute the Simes’ combined <span class="math inline">\(p\)</span>-value across all genes to reject the
doublet hypothesis for <span class="math inline">\(j'\)</span>. This
would provide a more rigorous approach to ruling out potential
doublet/parent combinations. However, this is very sensitive to
misspecification of clusters – see below.</p>
</div>
<div class="section level2">
<h2 id="calling-doublet-clusters">Calling doublet clusters<a class="anchor" aria-label="anchor" href="#calling-doublet-clusters"></a>
</h2>
<p>Assuming that most clusters are not comprised of doublets, we
identify clusters that have an unusually low number of detected genes
that violate the intermediacy condition. This is achieved by identifying
small outliers on the log-transformed number of detected genes, using
the median absolute deviation-based method in the function. (We use a
log-transformation simply to improve resolution at low values.) Clusters
are likely to be doublets if they are outliers on this metric.</p>
<p>Doublet clusters should also have larger library sizes than the
proposed parent clusters. This is consistent with the presence of more
RNA in each doublet, though the library size of the doublet cluster need
not be a sum of that of the parent clusters (due to factors such as
saturation and composition effects). The proportion of cells assigned to
the doublet cluster should also be “reasonable”; exactly what this means
depends on the experimental setup and the doublet rate of the protocol
in use.</p>
</div>
<div class="section level2">
<h2 id="discussion">Discussion<a class="anchor" aria-label="anchor" href="#discussion"></a>
</h2>
<p>The biggest advantage of this approach lies in its interpretability.
Given a set of existing clusters, we can explicitly identify those that
are likely to be doublets. We also gain some insight onto the parental
origins of each putative doublet cluster, which may be of some interest.
We avoid any assumptions about doublet formation that are otherwise
necessary for the simulation-based methods. In particular, we do not
require any knowledge about exact the relationship between <span class="math inline">\(s_j\)</span> and <span class="math inline">\(s_i\)</span>, allowing us to identify doublets
even when the exact location of the doublet is unknown (e.g., due to
differences in RNA content between the parent clusters).</p>
<p>The downside is that, of course, we are dependent on being supplied
with sensible clusters where the parental and doublet cells are
separated. The intermediacy requirement is loose enough to provide some
robustness against misspecification, but this only goes so far. In
addition, this strategy has a bias towards calling clusters with few
cells as doublets (or parents of doublets) because the DE detection
power is low. This can be somewhat offset by comparing
<code>num.de</code> against <code>median.de</code> as latter will be low
for clusters involved in systematically low-powered comparisons, though
it is difficult to adjust for the exact effect of the differences of
power on the IUT.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] scDblFinder_1.15.3          bluster_1.10.0             </span></span>
<span><span class="co">##  [3] scater_1.28.0               ggplot2_3.4.3              </span></span>
<span><span class="co">##  [5] scran_1.28.2                scuttle_1.10.2             </span></span>
<span><span class="co">##  [7] ensembldb_2.24.1            AnnotationFilter_1.24.0    </span></span>
<span><span class="co">##  [9] GenomicFeatures_1.52.2      AnnotationDbi_1.62.2       </span></span>
<span><span class="co">## [11] scRNAseq_2.14.0             SingleCellExperiment_1.22.0</span></span>
<span><span class="co">## [13] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">## [15] GenomicRanges_1.52.0        GenomeInfoDb_1.36.3        </span></span>
<span><span class="co">## [17] IRanges_2.34.1              S4Vectors_0.38.2           </span></span>
<span><span class="co">## [19] BiocGenerics_0.46.0         MatrixGenerics_1.12.3      </span></span>
<span><span class="co">## [21] matrixStats_1.0.0           BiocStyle_2.28.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.7                magrittr_2.0.3               </span></span>
<span><span class="co">##   [3] ggbeeswarm_0.7.2              farver_2.1.1                 </span></span>
<span><span class="co">##   [5] rmarkdown_2.25                fs_1.6.3                     </span></span>
<span><span class="co">##   [7] BiocIO_1.10.0                 zlibbioc_1.46.0              </span></span>
<span><span class="co">##   [9] ragg_1.2.5                    vctrs_0.6.3                  </span></span>
<span><span class="co">##  [11] memoise_2.0.1                 Rsamtools_2.16.0             </span></span>
<span><span class="co">##  [13] DelayedMatrixStats_1.22.6     RCurl_1.98-1.12              </span></span>
<span><span class="co">##  [15] htmltools_0.5.6               S4Arrays_1.0.6               </span></span>
<span><span class="co">##  [17] progress_1.2.2                AnnotationHub_3.8.0          </span></span>
<span><span class="co">##  [19] curl_5.0.2                    BiocNeighbors_1.18.0         </span></span>
<span><span class="co">##  [21] xgboost_1.7.5.1               sass_0.4.7                   </span></span>
<span><span class="co">##  [23] bslib_0.5.1                   desc_1.4.2                   </span></span>
<span><span class="co">##  [25] cachem_1.0.8                  GenomicAlignments_1.36.0     </span></span>
<span><span class="co">##  [27] igraph_1.5.1                  mime_0.12                    </span></span>
<span><span class="co">##  [29] lifecycle_1.0.3               pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [31] rsvd_1.0.5                    Matrix_1.5-4.1               </span></span>
<span><span class="co">##  [33] R6_2.5.1                      fastmap_1.1.1                </span></span>
<span><span class="co">##  [35] GenomeInfoDbData_1.2.10       shiny_1.7.5                  </span></span>
<span><span class="co">##  [37] digest_0.6.33                 colorspace_2.1-0             </span></span>
<span><span class="co">##  [39] rprojroot_2.0.3               dqrng_0.3.1                  </span></span>
<span><span class="co">##  [41] irlba_2.3.5.1                 ExperimentHub_2.8.1          </span></span>
<span><span class="co">##  [43] textshaping_0.3.6             RSQLite_2.3.1                </span></span>
<span><span class="co">##  [45] beachmat_2.16.0               labeling_0.4.3               </span></span>
<span><span class="co">##  [47] filelock_1.0.2                fansi_1.0.4                  </span></span>
<span><span class="co">##  [49] httr_1.4.7                    abind_1.4-5                  </span></span>
<span><span class="co">##  [51] compiler_4.3.1                bit64_4.0.5                  </span></span>
<span><span class="co">##  [53] withr_2.5.0                   BiocParallel_1.34.2          </span></span>
<span><span class="co">##  [55] viridis_0.6.4                 DBI_1.1.3                    </span></span>
<span><span class="co">##  [57] biomaRt_2.56.1                MASS_7.3-60                  </span></span>
<span><span class="co">##  [59] rappdirs_0.3.3                DelayedArray_0.26.7          </span></span>
<span><span class="co">##  [61] rjson_0.2.21                  tools_4.3.1                  </span></span>
<span><span class="co">##  [63] vipor_0.4.5                   beeswarm_0.4.0               </span></span>
<span><span class="co">##  [65] interactiveDisplayBase_1.38.0 httpuv_1.6.11                </span></span>
<span><span class="co">##  [67] glue_1.6.2                    restfulr_0.0.15              </span></span>
<span><span class="co">##  [69] promises_1.2.1                grid_4.3.1                   </span></span>
<span><span class="co">##  [71] Rtsne_0.16                    cluster_2.1.4                </span></span>
<span><span class="co">##  [73] generics_0.1.3                gtable_0.3.4                 </span></span>
<span><span class="co">##  [75] data.table_1.14.8             hms_1.1.3                    </span></span>
<span><span class="co">##  [77] metapod_1.8.0                 BiocSingular_1.16.0          </span></span>
<span><span class="co">##  [79] ScaledMatrix_1.8.1            xml2_1.3.5                   </span></span>
<span><span class="co">##  [81] utf8_1.2.3                    XVector_0.40.0               </span></span>
<span><span class="co">##  [83] ggrepel_0.9.3                 BiocVersion_3.17.1           </span></span>
<span><span class="co">##  [85] pillar_1.9.0                  stringr_1.5.0                </span></span>
<span><span class="co">##  [87] limma_3.56.2                  later_1.3.1                  </span></span>
<span><span class="co">##  [89] dplyr_1.1.3                   BiocFileCache_2.8.0          </span></span>
<span><span class="co">##  [91] lattice_0.21-8                rtracklayer_1.60.1           </span></span>
<span><span class="co">##  [93] bit_4.0.5                     tidyselect_1.2.0             </span></span>
<span><span class="co">##  [95] locfit_1.5-9.8                Biostrings_2.68.1            </span></span>
<span><span class="co">##  [97] knitr_1.44                    gridExtra_2.3                </span></span>
<span><span class="co">##  [99] bookdown_0.35                 ProtGenerics_1.32.0          </span></span>
<span><span class="co">## [101] edgeR_3.42.4                  xfun_0.40                    </span></span>
<span><span class="co">## [103] statmod_1.5.0                 stringi_1.7.12               </span></span>
<span><span class="co">## [105] lazyeval_0.2.2                yaml_2.3.7                   </span></span>
<span><span class="co">## [107] evaluate_0.21                 codetools_0.2-19             </span></span>
<span><span class="co">## [109] tibble_3.2.1                  BiocManager_1.30.22          </span></span>
<span><span class="co">## [111] cli_3.6.1                     xtable_1.8-4                 </span></span>
<span><span class="co">## [113] systemfonts_1.0.4             munsell_0.5.0                </span></span>
<span><span class="co">## [115] jquerylib_0.1.4               Rcpp_1.0.11                  </span></span>
<span><span class="co">## [117] dbplyr_2.3.3                  png_0.1-8                    </span></span>
<span><span class="co">## [119] XML_3.99-0.14                 parallel_4.3.1               </span></span>
<span><span class="co">## [121] ellipsis_0.3.2                pkgdown_2.0.7                </span></span>
<span><span class="co">## [123] blob_1.2.4                    prettyunits_1.1.1            </span></span>
<span><span class="co">## [125] sparseMatrixStats_1.12.2      bitops_1.0-7                 </span></span>
<span><span class="co">## [127] viridisLite_0.4.2             scales_1.2.1                 </span></span>
<span><span class="co">## [129] purrr_1.0.2                   crayon_1.5.2                 </span></span>
<span><span class="co">## [131] rlang_1.1.1                   KEGGREST_1.40.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Pierre-Luc Germain.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
